clusterloader2/pkg/measurement/common/apiserverlatency.go

```diff
@@ -0,0 +1,270 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package common
+
+import (
+	"fmt"
+	"sort"
+	"strconv"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/prometheus/common/model"
+	"k8s.io/apimachinery/pkg/util/sets"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/perf-tests/clusterloader2/pkg/measurement"
+	"k8s.io/perf-tests/clusterloader2/pkg/util"
+)
+
+const (
+	bigClusterNodeCountThreshold = 500
+	// We are setting 1s threshold for apicalls even in small clusters to avoid flakes.
+	// The problem is that if long GC is happening in small clusters (where we have e.g.
+	// 1-core master machines) and tests are pretty short, it may consume significant
+	// portion of CPU and basically stop all the real work.
+	// Increasing threshold to 1s is within our SLO and should solve this problem.
+	apiCallLatencyThreshold time.Duration = 1 * time.Second
+
+	// We use a higher threshold for list apicalls if the cluster is big (i.e having > 500 nodes)
+	// as list response sizes are bigger in general for big clusters. We also use a higher threshold
+	// for list calls at cluster scope (this includes non-namespaced and all-namespaced calls).
+	apiListCallLatencyThreshold      time.Duration = 5 * time.Second
+	apiClusterScopeListCallThreshold time.Duration = 10 * time.Second
+
+	currentApiCallMetricsVersion = "v1"
+)
+
+func init() {
+	measurement.Register("ApiServerLatency", createApiServerLatencyMeasurement)
+}
+
+func createApiServerLatencyMeasurement() measurement.Measurement {
+	return &apiServerLatencyMeasurement{}
+}
+
+type apiServerLatencyMeasurement struct{}
+
+// Execute supports two actions:
+// - reset - Resets latency data on api server side.
+// - gather - Gathers and prints current api server latency data.
+func (*apiServerLatencyMeasurement) Execute(config *measurement.MeasurementConfig) error {
+	action, err := util.GetString(config.Params, "action")
+	if err != nil {
+		return err
+	}
+
+	switch action {
+	case "reset":
+		glog.Infof("Resetting latency metrics in apiserver...")
+		body, err := config.ClientSet.CoreV1().RESTClient().Delete().AbsPath("/metrics").DoRaw()
+		if err != nil {
+			return err
+		}
+		if string(body) != "metrics reset\n" {
+			return fmt.Errorf("unexpected response: %q", string(body))
+		}
+		return nil
+	case "gather":
```

Comment:
Can you move this big code block below into separate function?

Reply:
done